opennebula/resource_opennebula_virtual_network.go

```diff
@@ -383,11 +397,15 @@ func resourceOpennebulaVirtualNetworkCreate(d *schema.ResourceData, meta interfa
 	controller := meta.(*goca.Controller)
 	var vnc *goca.VirtualNetworkController
 
+	rvnet := d.Get("reservation_vnet").(int)
+
 	// VNET reservation
-	if rvnet, ok := d.GetOk("reservation_vnet"); ok {
-		reservation_vnet := rvnet.(int)
+	if rvnet != -1 {
```

Comment:
I may miss something so I'm just asking, why did you do this ?

Reply:
Because GetOk returns false if the value is 0 and OpenNebula's indexes are zero based, so if you are using the first virtual network, this method returns a false negative. https://github.com/hashicorp/terraform-plugin-sdk/blob/6b1ff72b7c8fe08ca3a09a12f76e96c278ab1dc3/helper/schema/resource_data.go#L93

Reply:
oh yes, I shamefully forgot this

Reply:
I'm not sure what the reason behind that logic is, I was surprised to see it there as part of the SDK.

Reply:
There is a bunch of stuffs weird in the provider.
I think that due to old design choices. (not easy to find the exact information in the bunch of terraform issues, PR etc.)

As a side note, this provider still use the v1 of the SDK (I did the migration not long ago, it was the easy part). There is some room for improvement in migrating to the v2 of the SDK, I have some work pending to do this but I still have some problems to solve and it takes time, not sure I'll be able to do this soon.
Anyway I'm not sure that would solve the problem we are talking about but at least that would solve other problems.

Recently an other tool set appeared under the name "Plugin Framework", but it seems that we can't use it without rewriting the provider.

Here are some informations around the tools to write terraform providers:
https://www.terraform.io/docs/plugin/which-sdk.html